### ISSUE TITLE
audio dropout onResync

### ISSUE THREAD

**Issue by christf** [badaix/snapcast#433]
Even when snapserver is started onResync $significantly_smaller_value_than_buffersize leads to audio dropouts. The scenario happened today 37 times within 1h.

# Observations
* This (something similar?) was reported as #248 
* when placing mbuffer between mpd and snapcast (mbuffer -i /tmp/mpdoutput -m 10M -P 10 >/tmp/snapcastinput) I can see that snapcast starts producing audio data when mbuffer starts writing data to snapserver. The buffer fills up.
* I am using 44100:16:2 format in mpd. This means ~172 kbyte of data is generated per second by mpd. The amount with whith which the buffer fills up matches exactly the amount of data that is required to play audio for the amount of time reported by the onResync Error message.
* (after some experimentation I am not sure about this one, may have been coincidence: When re-starting snapserver, the onResync Times seem less frequent and much shorter (3ms after restart up to 6s before) on an otherwise unchanged system.)
* clientlog
```
2018-12-21 23-05-27 [Info] Chunk: -4    -3      0       10      425     223
2018-12-21 23-05-28 [Info] Chunk: -5    -4      0       10      433     226
2018-12-21 23-05-29 [Info] Chunk: -2    -4      -1      10      441     228
2018-12-21 23-05-30 [Info] Chunk: -1    -2      -1      10      449     228
2018-12-21 23-05-31 [Info] age > 200: -592
2018-12-21 23-05-31 [Info] sleep < -bufferDuration/2: -592 < -62, sleep: -588
2018-12-21 23-05-31 [Info] sleep < -bufferDuration/2: -588 < -62, sleep: -462
2018-12-21 23-05-31 [Info] sleep < -bufferDuration/2: -462 < -62, sleep: -342
2018-12-21 23-05-31 [Info] sleep < -bufferDuration/2: -342 < -62, sleep: -217
2018-12-21 23-05-31 [Info] sleep < -bufferDuration/2: -217 < -62, sleep: -92
2018-12-21 23-05-31 [Info] sleep < -bufferDuration/2: -92 < -62, sleep: 32
2018-12-21 23-05-31 [Info] Sleep 32, age: 36, bufferDuration: 124
2018-12-21 23-05-31 [Info] Chunk: 362   362     362     362     1       227
2018-12-21 23-05-31 [Info] Sleep 32, age: 32, bufferDuration: 124
2018-12-21 23-05-32 [Info] Chunk: 323   323     323     323     1       228
```
* serverlog:
```
2018-12-21 23-05-13 [Info] onStateChanged (default): 1
2018-12-21 23-05-13 [Info] onResync (default): 582ms
2018-12-21 23-05-13 [Info] onStateChanged (default): 2
```

# Conclusions
* Snapserver is not reading all the data that was provided. Buffer-underruns are the  consequence.
* This is completely unrelated to system load. This is a bug in snapserver.

I think the reason is the usage of sleep() in pipeStream.

**Comment by christf** [comment]
... testing a workaround that avoids the need to both: sleep and resync.

**Comment by christf** [comment]
I removed the patch from #432 again. While it reduces the amount of dropouts by factor >10 (5 dropouts in 2h music streaming compared to 37 in 1h) there is still an issue with stream opening/closing. Hitting stop in mpd and then re-starting the stream does not yet lead to audio being played. 

The patch will be prepared in https://github.com/christf/snapcast/tree/dropoutfix-server and raised as separate PR.

**Comment by badaix** [comment]
fixed in v0.19.0
